### PR TITLE
[OPIK-396] dynamically sort by last_updated_trace_at

### PR DESCRIPTION
**Pull request by idoberko2**
## Details
We want to achieve sorting by `last_updated_trace_at`. This field has a dynamically computed value and is not stored in MySQL. For that purpose we need to manually sort and paginate.

**Note:** For the time being we decided to not support sorting by multiple fields as that would introduce some messy logic.

## Issues
OPIK-396

## Testing
Added an E2E test that covers the new sorting logic. Pagination logic is covered with unit tests.

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java [hunk 1]
```diff
@@ -221,6 +228,10 @@ public Page<Project> find(int page, int size, @NonNull ProjectCriteria criteria,
 
         String workspaceId = requestContext.get().getWorkspaceId();
 
+        if (!sortingFields.isEmpty() && sortingFields.get(0).field().equals(SortableFields.LAST_UPDATED_TRACE_AT)) {
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java]:
```suggestion
        if (!sortingFields.isEmpty() && sortingFields.getFirst().field().equals(SortableFields.LAST_UPDATED_TRACE_AT)) {
```

File: apps/opik-backend/src/main/java/com/comet/opik/utils/PaginationUtils.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+
+public class PaginationUtils {
+    public static String ERR_PAGE_INVALID = "invalid value for page '%d'";
+    public static String ERR_SIZE_INVALID = "invalid value for size '%d'";
+
+    public static <T> List<T> paginate(int page, int size, @NonNull List<T> elements) {
+        if (page < 1) {
+            throw new IllegalArgumentException(ERR_PAGE_INVALID.formatted(page));
+        }
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/utils/PaginationUtils.java]: Preconditions.checkArgument

File: apps/opik-backend/src/main/java/com/comet/opik/utils/PaginationUtils.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        if (size < 1) {
+            throw new IllegalArgumentException(ERR_SIZE_INVALID.formatted(size));
+        }
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/utils/PaginationUtils.java]: Preconditions.checkArgument

Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/utils/PaginationUtils.java]: Preconditions.checkArgument

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java [hunk 4]
```diff
@@ -770,7 +772,96 @@ void getProjects__whenSortingProjectsByName__thenReturnProjectSortedByName(Direc
...
+            var actualProjects = actualEntity.content();
+            if (expected == Direction.DESC) {
+                for (int i = 0; i < NUM_OF_PROJECTS; i++) {
+                    assertThat(projects.get(NUM_OF_PROJECTS - i - 1).name()).isEqualTo(actualProjects.get(i).name());
+                }
+            } else {
+                for (int i = 0; i < NUM_OF_PROJECTS; i++) {
+                    assertThat(projects.get(i).name()).isEqualTo(actualProjects.get(i).name());
+                }
+            }
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]: Can we assert the whole list/object instead?

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java [hunk 5]
```diff
@@ -770,7 +772,96 @@ void getProjects__whenSortingProjectsByName__thenReturnProjectSortedByName(Direc
...
+            final int NUM_OF_PROJECTS = 5;
+            String workspaceName = UUID.randomUUID().toString();
+            String apiKey = UUID.randomUUID().toString();
+            String workspaceId = UUID.randomUUID().toString();
+
+            mockTargetWorkspace(apiKey, workspaceName, workspaceId);
+
+            List<Project> projects = IntStream.range(0, NUM_OF_PROJECTS)
+                    .mapToObj(i -> factory.manufacturePojo(Project.class))
+                    .toList();
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]:
```suggestion
            List<Project> projects = PodamFactoryUtils.manufacturePojoList(Project.class);
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java [hunk 6]
```diff
@@ -770,7 +772,96 @@ void getProjects__whenSortingProjectsByName__thenReturnProjectSortedByName(Direc
...
+                            StandardCharsets.UTF_8))
+                    .request()
+                    .header(HttpHeaders.AUTHORIZATION, apiKey)
+                    .header(WORKSPACE_HEADER, workspaceName)
+                    .get();
+
+            var actualEntity = actualResponse.readEntity(Project.ProjectPage.class);
+
+            assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(200);
+            assertThat(actualEntity.size()).isEqualTo(5);
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]: Also assert that page setup

Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]:
```suggestion
            assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(200);
            assertThat(actualEntity.size()).isEqualTo(projects.size());
```

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java [hunk 7]
```diff
@@ -253,6 +264,63 @@ public Page<Project> find(int page, int size, @NonNull ProjectCriteria criteria,
...
+    private List<UUID> sortByLastTrace(@NonNull Map<UUID, Instant> projectLastUpdatedTraceAtMap,
+            @NonNull SortingField sortingField) {
+        if (sortingField.direction() == Direction.DESC) {
+            return projectLastUpdatedTraceAtMap.entrySet().stream().sorted(reverseOrder(Map.Entry.comparingByValue()))
+                    .map((Map.Entry::getKey))
+                    .toList();
+        }
+
+        return projectLastUpdatedTraceAtMap.entrySet().stream().sorted(Map.Entry.comparingByValue())
+                .map((Map.Entry::getKey))
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java]: Minor
```suggestion
                .map(Map.Entry::getKey)
```

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java [hunk 8]
```diff
@@ -253,6 +264,63 @@ public Page<Project> find(int page, int size, @NonNull ProjectCriteria criteria,
...
+
+        return new ProjectPage(page, projects.size(), allProjectIds.size(), projects,
+                sortingFactory.getSortableFields());
+    }
+
+    private List<UUID> sortByLastTrace(@NonNull Map<UUID, Instant> projectLastUpdatedTraceAtMap,
+            @NonNull SortingField sortingField) {
+        if (sortingField.direction() == Direction.DESC) {
+            return projectLastUpdatedTraceAtMap.entrySet().stream().sorted(reverseOrder(Map.Entry.comparingByValue()))
+                    .map((Map.Entry::getKey))
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ProjectService.java]: Minor
```suggestion
                    .map(Map.Entry::getKey)
```

**Review by thiagohora — COMMENTED** [review]
Looks great, just some adjustments

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java (hunk 6 above)
Comment (idoberko2) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]: What do you mean by page setup?

Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]: like page, total, size and content

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java [hunk 9]
```diff
@@ -770,7 +772,96 @@ void getProjects__whenSortingProjectsByName__thenReturnProjectSortedByName(Direc
...
+            assertThat(actualEntity.size()).isEqualTo(projects.size());
+
+            var actualProjects = actualEntity.content();
+            if (expected == Direction.DESC) {
+                assertThat(actualProjects).usingRecursiveFieldByFieldElementComparatorIgnoringFields(IGNORED_FIELDS)
+                        .containsExactlyElementsOf(projects.reversed());
+            } else {
+                assertThat(actualProjects).usingRecursiveFieldByFieldElementComparatorIgnoringFields(IGNORED_FIELDS)
+                        .containsExactlyElementsOf(projects);
+            }
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectsResourceTest.java]:
```suggestion
var expectedProjects = expected == Direction.DESC? projects.reversed() : projects;

assertThat(actualProjects).usingRecursiveFieldByFieldElementComparatorIgnoringFields(IGNORED_FIELDS)
                        .containsExactlyElementsOf(expectedProjects);
```
```

**Review by thiagohora — DISMISSED** [review]

**Pull request merged**